R/rc_initialise.R

```diff
@@ -1,38 +1,363 @@
+    }
+ 
+    
+    # Set model parameters
+    parms <- list(unit = 'psomperfraction',
+                  initialization_method = 'none')
+    
+    # Set newly required inputs
+  
+    # Run initialization run for 150 years
+    this.result <- rc_sim(rothc_rotation = crop_extend,
+                          rothc_amendment = amendment_extend,
+                          soil_properties = soil_properties,
+                          weather = dt.weather,
+                          rothc_parms = parms)
+    
+    # take last two rotations
+    this.result.fin <- this.result[year > max(year)-2*nrow(crops),lapply(.SD,mean)]
   
-  # Prepare input for scenario Business As Usual
-  scen.inp <- rc_input_scenario(B_LU_BRP = B_LU_BRP, scen = 'BAU')
-  rotation <- scen.inp$rotation
-  amendment <- scen.inp$amendment
   
-  # Set model parameters
-  parms <- list(simyears = 150,unit = 'psomperfraction', initialize = TRUE)
+    fractions <- this.result.fin[,.(fr_IOM = CIOM / A_SOM_LOI,
+                                    fr_DPM = CDPM / A_SOM_LOI,
+                                    fr_RPM = CRPM / A_SOM_LOI,
+                                    fr_BIO = CBIO / A_SOM_LOI)]
+    
+  }
   
-  # Run initialization run for 150 years
-  this.result <- rc_sim(soil_properties = soil_properties,
-                               rothc_rotation = rotation,
-                               rothc_amendment = amendment,
-                               rothc_parms = parms)
+  ## calculate initial carbon pools assuming equilibrium in C pools, using analytical solution (Heuvelink)
+  if(initialization_method == 'spinup_analytical_heuvelink'){
+
+    # establish simyear based on time data table
+    isimyears <- max(dt.time$year) - min(dt.time$year) + 1
+    
+    # averaged total C input (kg C/ha/year) from crops and amendments
+    if (is.null(crops) || nrow(crops) == 0) {
+      c_input_crop <- 0
+    }else{
+    c_input_crop <- crops[,sum(B_C_OF_INPUT)/isimyears]
+    }
+    
+    if (is.null(amendment) || nrow(amendment) == 0) {
+       c_input_man <- 0
+    } else if (!is.null(amendment$B_C_OF_INPUT) && any(!is.na(amendment$B_C_OF_INPUT))) {
+        c_input_man <- amendment[, sum(B_C_OF_INPUT)/isimyears]
+    } else {
+          c_input_man <- amendment[, sum(P_DOSE * P_C_OF, na.rm = TRUE)/isimyears]
+    }
   
-  # take last two rotations
-  this.result.fin <- this.result[year > max(year)-2*nrow(rotation),lapply(.SD,mean)]
+
+    # calculate C input ratio of crop and amendments
+    c_input_tot <- c_input_crop+c_input_man
+    
+    if(c_input_tot <= 0 || !is.finite(c_input_tot)){
+      # if there are no C inputs, continue calculation based on crops only to prevent NaNs
+      CR_proportion <- 1
+      M_proportion <- 0
+    }else{
+    CR_proportion <- c_input_crop / (c_input_tot)
+    M_proportion <- c_input_man / (c_input_tot)
+    }
+    
+    # estimate DPM-RPM ratio of the inputs from crop and amendments
+    if (!is.null(crops) && nrow(crops) > 0) {
```

Comment:
i have the impression from some repetition here.
earlier you already did a check whether crops is there, same for amendment. 
if so, then within that if statement you can in principle also apply these calculations isn't?
that minimises the number of checks.